Java Driver: Connection is lost during changes(), never times out.

My code is essentially:

```java
try (Connection conn = dbp.getConnection()) {
  Cursor<HashMap<String, ?>> cursor = r.table("messages").changes().run(conn);
  for (HashMap<String, ?> item : cursor) {
    .. do stuff ...
  }
}
```

I connected to the server and was successfully listening to changes. Then, I killed the database server (only a single instance was running). As far as I can tell, the Java driver does not break the for loop when the connection is lost it just keeps sitting there expecting more data forever. There doesn't appear to be any timeout period. And, obviously, if I restart the database server then changes don't start coming in again on that old cursor.

So my question is: how are we supposed to detect connection losses and recover from them if the driver just sits there in the loop?

I can confirm that closing the connection manually breaks the loop. Program and logs attached.

```
17:45:18.160 [Thread-0] DEBUG d.SimpleTest - Thread A listening for changes...
17:45:18.228 [Thread-0] DEBUG c.r.a.Query - JSON Send: Token: 1 [1,[152,[[15,["messages"]]]],{"db":[14,["chatapp"]]}]
17:45:18.243 [pool-1-thread-1] DEBUG c.r.a.Query - JSON Recv: Token: 1 {"t":3,"r":[],"n":[1]}
17:45:18.373 [Thread-0] DEBUG c.r.a.Query - JSON Send: Token: 1 [2]
17:45:48.159 [Thread-1] DEBUG d.SimpleTest - Thread B: Inserting first change.
17:45:48.161 [Thread-1] DEBUG c.r.a.Query - JSON Send: Token: 2 [1,[56,[[15,["messages"]],[3,[],{"name":"Example","message":"Hello World! (1)","time":0}]]],{"db":[14,["chatapp"]]}]
17:45:48.169 [pool-1-thread-1] DEBUG c.r.a.Query - JSON Recv: Token: 1 {"t":3,"r":[{"new_val":{"id":"438e9428-2f91-45f8-b9cd-170556675f4f","message":"Hello World! (1)","name":"Example","time":0},"old_val":null}],"n":[1]}
17:45:48.175 [pool-1-thread-1] DEBUG c.r.a.Query - JSON Recv: Token: 2 {"t":1,"r":[{"deleted":0,"errors":0,"generated_keys":["438e9428-2f91-45f8-b9cd-170556675f4f"],"inserted":1,"replaced":0,"skipped":0,"unchanged":0}]}
17:45:48.176 [Thread-0] DEBUG d.SimpleTest - Thread A received change: {new_val={name=Example, id=438e9428-2f91-45f8-b9cd-170556675f4f, time=0, message=Hello World! (1)}, old_val=null}
17:45:48.176 [Thread-0] DEBUG c.r.a.Query - JSON Send: Token: 1 [2]
17:46:18.177 [Thread-1] DEBUG d.SimpleTest - Thread B: Inserting second change.
17:46:18.177 [Thread-1] DEBUG c.r.a.Query - JSON Send: Token: 3 [1,[56,[[15,["messages"]],[3,[],{"name":"Example","message":"Hello World! (2)","time":0}]]],{"db":[14,["chatapp"]]}]
17:46:18.179 [pool-1-thread-1] DEBUG c.r.a.Query - JSON Recv: Token: 1 {"t":3,"r":[{"new_val":{"id":"de0f224c-764a-4ea2-a22d-f1e10b0854cf","message":"Hello World! (2)","name":"Example","time":0},"old_val":null}],"n":[1]}
17:46:18.180 [Thread-0] DEBUG c.r.a.Query - JSON Send: Token: 1 [2]
17:46:18.180 [Thread-0] DEBUG d.SimpleTest - Thread A received change: {new_val={name=Example, id=de0f224c-764a-4ea2-a22d-f1e10b0854cf, time=0, message=Hello World! (2)}, old_val=null}
17:46:18.188 [pool-1-thread-1] DEBUG c.r.a.Query - JSON Recv: Token: 3 {"t":1,"r":[{"deleted":0,"errors":0,"generated_keys":["de0f224c-764a-4ea2-a22d-f1e10b0854cf"],"inserted":1,"replaced":0,"skipped":0,"unchanged":0}]}
17:46:48.189 [Thread-1] DEBUG d.SimpleTest - Thread B: Dropping connection.
17:46:48.199 [Thread-0] DEBUG d.SimpleTest - Thread A failed: 
com.rethinkdb.gen.exc.ReqlDriverError: java.util.concurrent.CancellationException
	at com.rethinkdb.net.Cursor.waitOnCursorItems(Cursor.java:102) ~[rethinkdb-driver-2.3.0.jar:na]
	at com.rethinkdb.net.Cursor$DefaultCursor.getNext(Cursor.java:186) ~[rethinkdb-driver-2.3.0.jar:na]
	at com.rethinkdb.net.Cursor.next(Cursor.java:124) ~[rethinkdb-driver-2.3.0.jar:na]
	at demo.SimpleTest2$1.run(SimpleTest2.java:38) ~[classes/:na]
Caused by: java.util.concurrent.CancellationException: null
	at java.util.concurrent.CompletableFuture.cancel(Unknown Source) ~[na:1.8.0_11]
	at com.rethinkdb.net.Connection.lambda$close$2(Connection.java:161) ~[rethinkdb-driver-2.3.0.jar:na]
	at com.rethinkdb.net.Connection$$Lambda$1/1784662007.accept(Unknown Source) ~[na:na]
	at java.util.concurrent.ConcurrentHashMap$ValueSpliterator.forEachRemaining(Unknown Source) ~[na:1.8.0_11]
	at java.util.stream.ReferencePipeline$Head.forEach(Unknown Source) ~[na:1.8.0_11]
	at com.rethinkdb.net.Connection.close(Connection.java:156) ~[rethinkdb-driver-2.3.0.jar:na]
	at com.rethinkdb.net.Connection.close(Connection.java:136) ~[rethinkdb-driver-2.3.0.jar:na]
	at demo.SimpleTest2$2.run(SimpleTest2.java:69) ~[classes/:na]
17:47:18.192 [Thread-1] DEBUG d.SimpleTest - Thread B: Quiting.
17:47:18.192 [main] DEBUG d.SimpleTest - Workers joined, exiting.
```

```java
package demo;

import static com.rethinkdb.RethinkDB.r;

import java.util.HashMap;
import java.util.concurrent.TimeUnit;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.rethinkdb.net.Connection;
import com.rethinkdb.net.Cursor;

/**
 * Run in the web UI to add a row and get notified:
 * r.db("chatapp").table("messages").insert({"name": "Bob", "time": 0, "message": "Hello!"})
 */
public class SimpleTest2 {
  private static final Logger logger = LoggerFactory.getLogger(SimpleTest.class);
  
  public static void main(String[] args) throws Exception {    
    final Connection conn = r.connection()
        .hostname("localhost")
        .port(28015)
        .db("chatapp")
        .connect();
    
    Thread A = new Thread() {
      @Override
      public void run() {
        try {
          Cursor<HashMap<String, ?>> cursor = null;
          
          try {
            logger.debug("Thread A listening for changes...");
            cursor = r.table("messages").changes().run(conn);
            
            for (HashMap<String, ?> item : cursor) {
              logger.debug("Thread A received change: {}", item);            
            }
          } finally {
            if (cursor != null) {
              cursor.close();
            }
          }
        } catch (Exception e) {
          logger.debug("Thread A failed: ", e);
        }
      }
    };
    
    Thread B = new Thread() {
      public void run() {
        try {
          Thread.sleep(TimeUnit.SECONDS.toMillis(30));
          logger.debug("Thread B: Inserting first change.");
          r.table("messages").insert(
              r.hashMap("name", "Example")
              .with("message", "Hello World! (1)")
              .with("time", 0)).run(conn);
          Thread.sleep(TimeUnit.SECONDS.toMillis(30));
          logger.debug("Thread B: Inserting second change.");
          r.table("messages").insert(
              r.hashMap("name", "Example")
              .with("message", "Hello World! (2)")
              .with("time", 0)).run(conn);
          Thread.sleep(TimeUnit.SECONDS.toMillis(30));
          logger.debug("Thread B: Dropping connection.");
          conn.close();
          Thread.sleep(TimeUnit.SECONDS.toMillis(30));
          logger.debug("Thread B: Quiting.");
        } catch (Exception e) {
          logger.debug("Thread B failed: ", e);
        }
      }
    };
    
    A.start();
    B.start();
    A.join();
    B.join();
    logger.debug("Workers joined, exiting.");
  }
}
```

--- a/drivers/java/src/main/java/com/rethinkdb/net/Connection.java
+++ b/drivers/java/src/main/java/com/rethinkdb/net/Connection.java
@@ -118,7 +118,7 @@ public class Connection implements Closeable {
                     if (awaiter != null) {
                         awaiter.complete(response);
                     }
-                } catch (IOException e) {
+                } catch (Exception e) {
                     awaiterException = e;
                     this.close();
                     break;

